### ISSUE TITLE
.query timeout

### ISSUE THREAD

**Issue by 4ldas** [stevepryde/thirtyfour#122]
It seems like it times out way too fast for me, i saw an earlier issue for .set_timeout_poller() on web driver however it has depreciated since then or something since i couldnt find the method. So how would i do it now and id appreciate an example so i guess just so i can be clear

**Comment by stevepryde** [comment]
If you just want to increase the timeout, you can do it per-call using `ElementQuery::wait()`
https://docs.rs/thirtyfour/latest/thirtyfour/extensions/query/struct.ElementQuery.html#method.wait

If you want it to be the default, you need to construct the `ElementQuery` yourself and specify your own instance of `ElementPollerWithTimeout`.

You could create your own helper for this as follows:
```rust
use thirtyfour::extensions::query::{ElementPollerWithTimeout, ElementQuery};

fn my_query(driver: &WebDriver, by: By) -> ElementQuery {
    let poller = ElementPollerWithTimeout::new(Duration::from_secs(60), Duration::from_millis(500));
    driver.query(by).with_poller(Box::new(poller))
}
```

If you want it as a method on `WebDriver` you can create a trait with the above method and implement it for `WebDriver`.

The reason this all changed was because I was trying to remove the need for `WebDriver` to contain any state. I still think this is the right move because maintaining state across the entire `WebDriver` session seems like it could cause issues, much like global variables. One way around this would be to create your own customized wrapper around `WebDriver` with its own `query()` method (implemented as above), and implement `Deref` to expose all the existing `WebDriver` functionality as well. If you have other ideas I'd be happy to consider them.

**Comment by 4ldas** [comment]
I have it as a trait function for a type and im trying to call it from main function and it feels as if theres like some webdriver timeout it expects by default or something i got no clue since i tried using both the wait and adding the poller and neither worked. It still timer me out, the error btw 
```bash
thread 'main' panicked at "called `Result::unwrap()` on an `Err` value: timeout: ",  src/main.rs 16:32```

**Comment by 4ldas** [comment]
> I have it as a trait function for a type and im trying to call it from main function and it feels as if theres like some webdriver timeout it expects by default or something i got no clue since i tried using both the wait and adding the poller and neither worked. It still timer me out, the error btw 
> ```bash
> thread 'main' panicked at "called `Result::unwrap()` on an `Err` value: timeout: ",  src/main.rs 16:32
> ```

Or maybe this is async issue how would i do this without async?

**Comment by stevepryde** [comment]
What you're describing is consistent with a query timing out. Can you post the code that is running into this error?

**Comment by 4ldas** [comment]
> What you're describing is consistent with a query timing out. Can you post the code that is running into this error?

Dont wanna share exact code, however i do have a struct and implimenting a function for it which accepts arguments of itself and the webdriver. This function is called from the main function and goes over to my desired website and queries for either element 1 or 2 if first one is not found. The website is quite slow thus causing the time out. The driver is initialized in main function and passed over to the function and after gets called to quit in the main function

**Comment by stevepryde** [comment]
The timeout message is likely to be from an element wait (i.e. waiting for some element condition via `ElementWaiter`), or perhaps while waiting for the page to load. It might not be the query at all.

Also check the page load timeout setting.

**Comment by 4ldas** [comment]
> The timeout message is likely to be from an element wait (i.e. waiting for some element condition via `ElementWaiter`), or perhaps while waiting for the page to load. It might not be the query at all.
> 
> Also check the page load timeout setting.

Makes sense now, im waiting until that query is displayed with .until_displayed() so i know when my code should execute, how would i call wait on that i guess, i would assume it would be with .query but nevermind maybe

**Comment by stevepryde** [comment]
Btw you can replace the poller in `ElementWaiter` the same as you can in `ElementQuery` as well as changing the wait params per request too.

**Comment by 4ldas** [comment]
> Btw you can replace the poller in `ElementWaiter` the same as you can in `ElementQuery` as well as changing the wait params per request too.

I tried adding .wait to both and still issue. To be clear i do 
```
.wait_until().displayed().await?;
```
on query, thats without the wait added well

**Comment by stevepryde** [comment]
I'll take a look on the weekend and give you some example code.

By default the poller will poll every half second for up to 30 seconds. In this case it's waiting for an element to become displayed. If the element never gets displayed this will return the timeout you were seeing. It's worth checking that the element is the correct one (i.e your selector did select the correct element). And you can show the devtools in the browser and make sure the element is eventually displayed.

**Comment by 4ldas** [comment]
> I'll take a look on the weekend and give you some example code.
> 
> By default the poller will poll every half second for up to 30 seconds. In this case it's waiting for an element to become displayed. If the element never gets displayed this will return the timeout you were seeing. It's worth checking that the element is the correct one (i.e your selector did select the correct element). And you can show the devtools in the browser and make sure the element is eventually displayed.

It does get displayed however in appr 20 seconds only

**Comment by stevepryde** [comment]
It's also worth mentioning that instead of querying an element and then waiting for it to be displayed, it's probably generally better to just query the element with the "displayed()" filter on the query.

**Comment by stevepryde** [comment]
After playing with a small example I decided that using a custom poller was more cumbersome than I would like. So I added `WebDriverConfig` back, which allows setting the default poller for a `WebDriver` instance and all derived elements.

I've published `v0.32.0-rc.3` which includes this new functionality. See the new custom poller example in `examples/query/custom_poller.rs`.

Let me know if this helps to solve your issue.
